extmod/modframebuf.c

```diff
@@ -97,13 +97,69 @@ STATIC void rgb565_fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, i
     }
 }
 
+// Functions for GRAY4 format
+
+STATIC void gray4_setpixel(const mp_obj_framebuf_t *fb, int x, int y, uint32_t col) {
+    uint8_t *pixel = &((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1];
+
+    if (x % 2) {
+        *pixel = ((uint8_t)col & 0x0F) | (*pixel & 0xF0);
+    }
+    else {
+        *pixel = ((uint8_t)col << 4) | (0x0F & *pixel);
+    }
+}
+
+STATIC uint32_t gray4_getpixel(const mp_obj_framebuf_t *fb, int x, int y) {
+    if (x % 2) {
+        return ((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1] & 0x0F;
+    }
+
+    return ((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1] >> 4;
+}
+
+STATIC void gray4_fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, int h, uint32_t col) {
+    uint8_t *pixel_pair = &((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1];
+    uint8_t col_cleaned = (uint8_t)col & 0x0F;
+    uint8_t col_shifted_left = col_cleaned << 4;
+    uint8_t colored_pixel_pair = col_shifted_left | col_cleaned;
```

Comment:
First there was a for-loop instead of memset() and it had larger impact on speed. But it's still a good idea to calculate values (that will not change inside loops) outside. No unneeded memory allocations, no unneeded cpu usage aswell. Drawing graphics is not a strong part especially in case of I2C/SPI + bigger screen resolution. So I think it's better to do drawing as fast as possible everywhere it is possible. If we don't want a slideshow instead of animation.

And again - readability )

the only thing to optimize by removing is col_cleaned indeed.